src/v/datalake/translation/scheduling_policies.cc

```diff
@@ -317,4 +397,129 @@ ss::future<> fair_scheduling_policy::on_resource_exhaustion(
     }
 }
 
+std::optional<fair_scheduling_policy::finish_choice_info>
+fair_scheduling_policy::choose_translator_to_finish(executor& executor) {
+    auto& translators_to_finish = executor.translators_for_immediate_finish;
+    if (translators_to_finish.empty()) {
+        return std::nullopt;
+    }
+
+    // choice: the return value
+    //
+    // finish_key: index into `translators_to_finish` so that we can remove the
+    // final choice. we can't hold on to an iterator because the set is modified
+    // during iteration.
+    struct choice_info {
+        finish_choice_info info;
+        size_t finish_key;
+    };
+    std::optional<choice_info> choice;
+
+    // find highest priority running translator
+    for (auto it = translators_to_finish.begin();
+         it != translators_to_finish.end();) {
+        // map translator_id -> translator_executable
+        auto eit = executor.translators.find(it->second);
+        if (eit == executor.translators.end()) {
+            it = translators_to_finish.erase(it);
+            continue;
+        }
+
+        // status of translator of the current loop iteration
+        const auto curr_status = [eit] {
+            if (eit->second._running_hook.is_linked()) {
+                return finish_choice_info::status::running;
+            } else if (eit->second._waiting_hook.is_linked()) {
+                return finish_choice_info::status::waiting;
+            }
+            return finish_choice_info::status::idle;
+        }();
+
+        // potential return value for choice
+        const auto curr_info = finish_choice_info{
+          .it = eit,
+          .status = curr_status,
+        };
+
+        // first choice is a running translator
+        if (curr_info.status == finish_choice_info::status::running) {
+            choice = choice_info{.info = curr_info, .finish_key = it->first};
+            break;
+        }
+
+        if (!choice.has_value()) {
+            choice = choice_info{.info = curr_info, .finish_key = it->first};
+            // continue looking for a running translator
+        }
+
+        ++it;
+    }
+
+    if (choice.has_value()) {
+        translators_to_finish.erase(choice->finish_key);
+        return choice->info;
+    }
+
+    return std::nullopt;
+}
+
+ss::future<> fair_scheduling_policy::finish_translator(
+  executor& executor, finish_choice_info choice) {
+    auto& executable = choice.it->second;
+    executable.translator_ptr()->set_finish_translation();
```

Comment:
What guarantees us that the iterator is still valid when we get here (e.g. what if the partition moved)? Should we just store the translator id in `choice` and do the lookup again? Same question for down below when we're polling

Reply:
yep, you're right. it was guaranteed earlier when we were doing one-at-a-time finishing because there were no scheduling points, but that doesn't hold with running them in parallel.

fixed

Reply:
> below when we're polling

down there (if are talking about the same location) we do look-up the translator again for precisely the same reason you highlight. are we talking about the same location and the same situation?

Reply:
Ah you're right, I was thinking `it` was still `choice.it`, sorry!

Reply:
i replaced `choice.it` with `choice.id` and just repeat the lookups when necessary in case the translators peace out unexpectedly